draft-ietf-httpbis-connect-tcp.md

```diff
@@ -145,11 +145,13 @@ Authentication to a templated TCP proxy normally uses ordinary HTTP authenticati
 
 Clients SHOULD assume that all proxy resources generated by a single template share a protection space (i.e., a realm) ({{?RFC9110, Section 11.5}}).  For many authentication schemes, this will allow the client to avoid waiting for a "401 (Unauthorized)" response before each new connection through the proxy.
 
-## Relationship to the Capsule Protocol
+### Use of the Capsule Protocol
 
-Unlike the datagram-oriented templated HTTP proxying specifications {{CONNECT-UDP}}{{CONNECT-IP}}, this specification does not make use of the Capsule Protocol {{?RFC9297}}.  A future specification could define a procedure for performing TCP proxying using the Capsule Protocol, but no such procedure is defined here.
+In this specification, use of the Capsule Protocol {{?RFC9297}} is OPTIONAL.  Clients MAY request use of the Capsule Protocol by including a "Capsule-Protocol: ?1" header field in the request, and servers MAY confirm its use by echoing this header field in the response.
 
-When implementing this specification, clients and servers MUST NOT send a "Capsule-Protocol: ?1" header field.
+When using the Capsule Protocol, TCP data is sent using a new Capsule Type named DATA ({{data-capsule}}).  The DATA type is only applicable to Extended CONNECT protocols for which the use of the Capsule Protocol is optional.  It conveys a payload that should be processed in the same way as ordinary stream contents when the Capsule Protocol is not in use.
+
+A client that sends this header field MUST NOT send any data on the stream until it receives a response, because it does not know whether the server will treat the data as Capsule Protocol or TCP contents.
```

Comment:
This latency cost is unfortunate, and could be avoided by making Capsule Protocol mandatory to implement for servers.  However, that would increase the implementation complexity for a minimal compliant server, and should not be done lightly.

@DavidSchinazi @mnot

Reply:
Yeah, current design allows you to do various optimisations such calling splice (2) to avoid making copies to / from userspace, or pass the socket to some other process (with TLS being offloaded to the kernel). If we are to mandate the use of Caspules, those optimisations become impossible.

Reply:
A middle ground would be to keep this text, and later define a way for servers to commit to supporting the Capsule Protocol.  Note that the HTTPS record would likely not work for this commitment, because it must be fully authenticated.

Reply:
I share the performance concerns. Something Ilthat came up when I talked about this with @DavidSchinazi is the possibility of having a $humongous data frame that extends to the end of the stream, so that endpoints can opt out of needing capsule framing if they don't believe they need to send any other capsules.

I havent thought exactly about how that works across intermediaries for the different use cases we discussed during the IETF 120 neeting

Reply:
@LPardue 
> Something Ilthat came up when I talked about this with @DavidSchinazi is the possibility of having a $humongous data frame that extends to the end of the stream, so that endpoints can opt out of needing capsule framing if they don't believe they need to send any other capsules.

I'm not sure if that solves the performance concerns; the approach provides a escape hatch for senders, but it does not provide a way for the receivers to stay out from the business of decoding Capsules.

Reply:
> I really do care about allowing the client to send data with the request.

OK, I've reworked this text to make that safe while keeping capsule support optional, based on your suggestion.

> there is no reason to optimize HTTP/1.1 at this point.

I have heard that HTTP/3->HTTP/1.1 conversion at a gateway is a common pattern, so the performance of HTTP/1.1 may be significant even if it is not heavily used over the open internet.